### PR TITLE
build(docker): Optimize Docker image layers/caching & bump node to 14.15

### PR DESCRIPTION
**Pull request by TheCatLady**
#### Description
Since each `RUN` command in `Dockerfile` creates an additional layer, we can combine the installation of `tzdata` and writing of `committag.json` to avoid an unnecessary extra image layer.  🙂

The following `COPY` statements may also be combinable, but I'm not sure if there's anything in `BUILD_IMAGE` which needs to be excluded... so someone else will have to provide some input on that.  Is there a reason why we can't just copy `/app` from `BUILD_IMAGE`?  Could we clean up the `/app` directory in the first stage so we only need one `COPY` command here?

```
COPY . /app
WORKDIR /app

# copy from build image
COPY --from=BUILD_IMAGE /app/dist ./dist
COPY --from=BUILD_IMAGE /app/.next ./.next
COPY --from=BUILD_IMAGE /app/node_modules ./node_modules
```

#### Screenshot (if UI related)
N/A

#### Todos

- [x] Sucessfully builds `docker build`

### CODE REVIEW

**Comment by ankarhem** [comment]
While this is true, you will no longer cache the installation of the package, making it run each time. I think we want to optimize build time and not the (small?) difference it makes to have them share image layer.

**Comment by TheCatLady** [comment]
> While this is true, you will no longer cache the installation of the package, making it run each time. I think we want to optimize build time and not the (small?) difference it makes to have them share image layer.

The `tzdata` package is only 431.7 kB, however (3.36 MB installed).  It would only be a difference of a few seconds, no?

**Comment by ankarhem** [comment]
Yeah, but not cacheing it would create a new layer that includes that space for every new build. Instead of just once? So I don't think that's an argument that would be in favor for this PR? Now that I think of it's probably not cached now anyway since commit tag is over and would be changing?

File: Dockerfile [hunk 1]
```diff
@@ -11,22 +11,25 @@ RUN yarn --frozen-lockfile && \
...
 
+RUN apk add --no-cache tzdata
+
 ARG COMMIT_TAG
 ENV COMMIT_TAG=${COMMIT_TAG}
 
-RUN apk add tzdata
-
-COPY . /app
+COPY [^src]* /app/
```
Comment (sct) [inline Dockerfile:28]: Whats the point of doing this if we don't do it for the `server` directory also?

File: Dockerfile [hunk 2]
```diff
@@ -11,24 +11,19 @@ RUN yarn --frozen-lockfile && \
...
 RUN yarn install --production --ignore-scripts --prefer-offline
-RUN yarn cache clean
 
-FROM node:12.18-alpine
+RUN echo "{\"commitTag\": \"${COMMIT_TAG}\"}" > committag.json && \
+  rm -rf src && \
+  rm -rf server
 
-ARG COMMIT_TAG
-ENV COMMIT_TAG=${COMMIT_TAG}
```
Comment (sct) [inline Dockerfile:19]: We can't remove these.

File: Dockerfile [hunk 3]
```diff
@@ -11,24 +11,19 @@ RUN yarn --frozen-lockfile && \
 
 # remove development dependencies
 RUN yarn install --production --ignore-scripts --prefer-offline
-RUN yarn cache clean
 
-FROM node:12.18-alpine
+RUN echo "{\"commitTag\": \"${COMMIT_TAG}\"}" > committag.json && \
+  rm -rf src && \
+  rm -rf server
```
Comment (sct) [inline Dockerfile:17]: I don't like the idea of combining two entirely different operations into a single step.

Comment (TheCatLady) [reply Dockerfile]: Split into two `RUN` statements now!

File: Dockerfile (hunk 2 above)
Comment (TheCatLady) [reply Dockerfile:19]: Ah okay, added them back.

**Review by sct — APPROVED** [review]

**Pull request merged**